### PR TITLE
fixes #5208: Gradle build fails with android studio flamingo update

### LINKED ISSUE #5208: Gradle build fails with Android Studio Flamingo update
### Summary

Gradle build fails after updating to the latest Android Studio stable release (Flamingo)

**Why does it fail?**

Android studio Flamingo comes [bundled with JDK 17](https://developer.android.com/studio/releases#jdk-17). The minimum Gradle version to support JDK 17 is Gradle 7.3. But the project currently uses Gradle 6.9.

image

After upgrading Gradle from 6.9 to 7.3, the build still fails with the following error.

image

This is caused due to error in the build.gradle file of the data-client ie: [wikimedia-android-data-client](https://github.com/wikimedia/wikimedia-android-data-client) library, rather than the build.gradle of the commons app.

### Steps to reproduce

1. Upgrade Android Studio to Flamingo
2. Open **apps-android-commons** project
3. The build fails with the said error

### Expected behaviour

The build error is expected as the Gradle version is incompatible with the JDK version.

### Actual behaviour

NA

### Device name

Realme 3 pro

### Android version

Android 11

### Commons app version

4.0.3-debug

### Device logs

_No response_

### Screen-shots

_No response_

### Would you like to work on the issue?

Yes

### PR DESCRIPTION
**Pull request by siva-subramaniam-v**
Fixes #5208 

**What changes did you make and why?**

1. Upgraded Gradle version from 6.9 to 8.1.1
2. Upgraded AGP version from 4.0.1 to 8.0.0

**Tested betaDebug on**
1. Realme 3 Pro with API level 30 (Android 11)

Individual comments have been added to specific changes. Changes that do not contain individual comments are either self-explanatory or made by the AGP upgrade assistant.

**Note:** As part of this upgrade a [change](https://gerrit.wikimedia.org/r/c/translatewiki/+/913941/3) was required in [translatewiki](https://github.com/wikimedia/translatewiki) which has been merged.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -168,7 +168,7 @@ project.gradle.taskGraph.whenReady {
 }
 
 android {
-    compileSdkVersion 31
+    compileSdkVersion 33
```
Comment (siva-subramaniam-v) [inline app/build.gradle:171]: Changed CompileSdkVersion from 31 to 33. Fixes error as described in this [stack overflow issue](https://stackoverflow.com/questions/69943549/one-or-more-issues-found-when-checking-aar-metadata-values).

Comment (siva-subramaniam-v) [inline app/build.gradle:171]: Changed CompileSdkVersion from 31 to 33. Fixes error as described in this [stack overflow issue](https://stackoverflow.com/questions/69943549/one-or-more-issues-found-when-checking-aar-metadata-values).

File: app/build.gradle [hunk 2]
```diff
@@ -20,8 +20,8 @@ dependencies {
     // Utils
     implementation 'in.yuvi:http.fluent:1.3'
     implementation 'com.google.code.gson:gson:2.8.5'
-    implementation ("com.squareup.okhttp3:okhttp:$OKHTTP_VERSION"){
-        force = true //API 19 support
+    implementation ("com.squareup.okhttp3:okhttp:$OKHTTP_VERSION!!"){
+        //force = true //API 19 support
```
Comment (siva-subramaniam-v) [inline app/build.gradle:24]: Forcing dependency versions using force = true on a first-level dependency has been deprecated.

Refer to [Gradle documentation](https://docs.gradle.org/7.5/userguide/upgrading_version_5.html#forced_dependencies).

File: app/build.gradle [hunk 3]
```diff
@@ -341,7 +340,11 @@ android {
     buildFeatures {
         viewBinding true
     }
-
+    namespace 'fr.free.nrw.commons'
```
Comment (siva-subramaniam-v) [inline app/build.gradle:343]: Added the namespace DSL to the android {} block in the module-level build.gradle file and removed the package attribute in the manifest.

This change was made as part of preparing the app build for the Android Studio Flamingo release.

Refer to this [medium article](https://medium.com/androiddevelopers/5-ways-to-prepare-your-app-build-for-android-studio-flamingo-release-da34616bb946).

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    xmlns:tools="http://schemas.android.com/tools"
-    package="fr.free.nrw.commons">
```
Comment (siva-subramaniam-v) [inline app/src/main/AndroidManifest.xml:4]: Refer to[ this comment](https://github.com/commons-app/apps-android-commons/pull/5220/files#r1183287416).

File: app/src/main/java/fr/free/nrw/commons/customselector/helper/OnSwipeTouchListener.kt [hunk 5]
```diff
@@ -16,7 +16,7 @@ open class OnSwipeTouchListener(context: Context?) : View.OnTouchListener {
     private val SWIPE_THRESHOLD_WIDTH = (getScreenResolution(context!!)).first / 3
     private val SWIPE_VELOCITY_THRESHOLD = 1000
 
-    override fun onTouch(view: View?, motionEvent: MotionEvent?): Boolean {
+    override fun onTouch(view: View?, motionEvent: MotionEvent): Boolean {
```
Comment (siva-subramaniam-v) [inline app/src/main/java/fr/free/nrw/commons/customselector/helper/OnSwipeTouchListener.kt:19]: Changed parameter type from motionEvent: MotionEvent? to MotionEvent.

Fixes error: `Type mismatch: inferred type is MotionEvent? but MotionEvent was expected`

File: app/src/main/java/fr/free/nrw/commons/customselector/helper/OnSwipeTouchListener.kt [hunk 6]
```diff
@@ -32,7 +32,7 @@ open class OnSwipeTouchListener(context: Context?) : View.OnTouchListener {
 
     inner class GestureListener : GestureDetector.SimpleOnGestureListener() {
 
-        override fun onDown(e: MotionEvent?): Boolean {
+        override fun onDown(e: MotionEvent): Boolean {
```
Comment (siva-subramaniam-v) [inline app/src/main/java/fr/free/nrw/commons/customselector/helper/OnSwipeTouchListener.kt:35]: Changed parameter type from e: MotionEvent? to MotionEvent.

Fixes error: `'onDown' overrides nothing`

File: build.gradle [hunk 7]
```diff
@@ -2,23 +2,23 @@
...
 }
 
 allprojects {
-  repositories {
-    google()
-    jcenter()
-    maven { url "https://jitpack.io" }
-    maven { url "https://maven.google.com" }
-  }
+    repositories {
```
Comment (siva-subramaniam-v) [inline build.gradle:17]: Replaced [deprecated](https://docs.gradle.org/7.5/userguide/upgrading_version_6.html#jcenter_deprecation) jcenter() convenience method with gradlePluginPortal() as JCenter was [shutdown](https://blog.gradle.org/jcenter-shutdown).

Gradle Plugin Portal was added as the JCenter dependency [CircleProgressBar 1.1.1](https://mvnrepository.com/artifact/com.dinuscxj/circleprogressbar/1.1.1) was not available on Maven Central.

Refer to this [stack overflow issue](https://stackoverflow.com/questions/66651640/jcenter-deprecation-impact-on-gradle-and-android) and this [github issue](https://github.com/gradle/gradle/issues/4020).

File: build.gradle [hunk 8]
```diff
@@ -2,23 +2,23 @@
...
     }
 }
 
 allprojects {
-  repositories {
-    google()
-    jcenter()
-    maven { url "https://jitpack.io" }
-    maven { url "https://maven.google.com" }
-  }
```
Comment (siva-subramaniam-v) [inline build.gradle:23]: Replaced [deprecated](https://docs.gradle.org/7.5/userguide/upgrading_version_6.html#jcenter_deprecation) jcenter() convenience method with gradlePluginPortal() as JCenter was [shutdown](https://blog.gradle.org/jcenter-shutdown).

Gradle Plugin Portal was added as the JCenter dependency [CircleProgressBar 1.1.1](https://mvnrepository.com/artifact/com.dinuscxj/circleprogressbar/1.1.1) was not available on Maven Central.

Refer to this [stack overflow issue](https://stackoverflow.com/questions/66651640/jcenter-deprecation-impact-on-gradle-and-android) and this [github issue](https://github.com/gradle/gradle/issues/4020).

File: build.gradle [hunk 9]
```diff
@@ -2,23 +2,23 @@
...
+    repositories {
+        google()
+        mavenCentral()
+        gradlePluginPortal() // potential jcenter() replacement
+        maven { url "https://jitpack.io" }
+        maven { url "https://maven.google.com" }
+    }
```
Comment (siva-subramaniam-v) [inline build.gradle:23]: Replaced [deprecated](https://docs.gradle.org/7.5/userguide/upgrading_version_6.html#jcenter_deprecation) jcenter() convenience method with gradlePluginPortal() as JCenter was [shutdown](https://blog.gradle.org/jcenter-shutdown).

Gradle Plugin Portal was added as the JCenter dependency [CircleProgressBar 1.1.1](https://mvnrepository.com/artifact/com.dinuscxj/circleprogressbar/1.1.1) was not available on Maven Central.

Refer to this [stack overflow issue](https://stackoverflow.com/questions/66651640/jcenter-deprecation-impact-on-gradle-and-android) and this [github issue](https://github.com/gradle/gradle/issues/4020).

File: data-client/build.gradle [hunk 10]
```diff
@@ -55,6 +62,7 @@ android {
         sourceCompatibility = '1.8'
         targetCompatibility = '1.8'
     }
+    namespace 'org.wikipedia.dataclient'
```
Comment (siva-subramaniam-v) [inline data-client/build.gradle:65]: Added the namespace DSL to the android {} block in the module-level build.gradle file and removed the package attribute in the manifest.

This change was made as part of preparing the app build for the Android Studio Flamingo release.

Refer to this [medium article](https://medium.com/androiddevelopers/5-ways-to-prepare-your-app-build-for-android-studio-flamingo-release-da34616bb946).

File: data-client/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -1,4 +1,4 @@
-<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="org.wikipedia.dataclient">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (siva-subramaniam-v) [inline data-client/src/main/AndroidManifest.xml:1]: Refer to this [comment](https://github.com/commons-app/apps-android-commons/pull/5220/files#r1183342136).

File: gradle.properties [hunk 12]
```diff
@@ -33,5 +32,7 @@ systemProp.http.proxyPort=0
 systemProp.http.proxyHost=
 android.useAndroidX=true
 android.enableJetifier=true
-android.enableR8=false
-android.jetifier.blacklist=bcprov-jdk15on
\ No newline at end of file
+android.jetifier.ignorelist=bcprov-jdk15on
```
Comment (siva-subramaniam-v) [inline gradle.properties:35]: Changed **android.jetifier.blacklist** to **android.jetifier.ignorelist**.

Refer to this GitHub [comment](https://github.com/cashapp/paparazzi/issues/245#issue-975739503).

File: data-client/build.gradle
Comment (siva-subramaniam-v) [inline data-client/build.gradle:1]: Removed code for publishing library that caused a few errors. Library publishing code is unnecessary as the library is discontinued and contained within the main project.

**Review by nicolas-raoul — APPROVED** [review]
The app is building and running fine with this PR merged.
I will test a bit more and merge if @sivaraam and @RitikaPahwa4444 are OK.

File: app/build.gradle (hunk 2 above)
Comment (sivaraam) [reply app/build.gradle:24]: > Also, the minSdkVersion is 29 for this project.

We've pushed it back to 21, recently. I hope that doesn't affect this change 🙂

File: gradle.properties [hunk 13]
```diff
@@ -14,7 +14,6 @@
 # org.gradle.parallel=true
 #Thu Mar 01 15:28:48 IST 2018
 org.gradle.jvmargs=-Xmx1536M
-android.enableBuildCache=true
```
[the remark this answers is not in the document]
Comment (sivaraam) [reply gradle.properties:17]: > [Gradle build cache](https://docs.gradle.org/current/userguide/build_cache.html) has [superseded](https://stackoverflow.com/questions/68703442/gradle-build-cache-superseded-build-cache) Android-specific build caches.

Do we not need to add `org.gradle.caching=true` in our gradle.properties to enable Gradle build cache ? 🤔

File: build.gradle [hunk 14]
```diff
@@ -2,23 +2,23 @@
 buildscript {
     repositories {
         google()
-        jcenter()
         mavenCentral()
         maven { url "https://plugins.gradle.org/m2/" }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.1'
+        classpath 'com.android.tools.build:gradle:8.0.0'
```
Comment (sivaraam) [inline build.gradle:9]: Looks like 8.0.2 is already out. Could you kindly bump this up to 8.0.2 ?

**Review by sivaraam — COMMENTED** [review]
Thanks for taking this up and making the app works on newer Android Studio versions!

For the note, the version 8 of Android Gradle plugin seems to bump the JDK requirement to 17. Mentioning it here just for reference. In case anyone is stuck with Java 11 or so, they could bump the Gradle plugin down to 7.3.1. The app seems to build fine with that too. Also that Gradle plugin version enables the building the app on Android Studio Dolphin and also likely on the Electric Eel version.

**Comment by sivaraam** [comment]
It also seems like the GitHub workflow would need to be tweaked a bit to start using JDK 17 to handle the changes in this MR. Could you kindly adjust that too @siva-subramaniam-v ?

**Comment by siva-subramaniam-v** [comment]
@sivaraam I tried making the CI work, but it doesn't seem to be working. I don't have much experience working with CI. Please assign it to someone with prior experience.

**Comment by sivaraam** [comment]
> @sivaraam I tried making the CI work, but it doesn't seem to be working. I don't have much experience working with CI. Please assign it to someone with prior experience.

I've tried putting in a change which should hopefully fix the pipeline. Let's see 🤞🏼

**Comment by nicolas-raoul** [comment]
```
fr.free.nrw.commons.AboutActivityUnitTests > checkActivityNotNull FAILED
    java.lang.NoSuchMethodError at ProxyMaker.java:111
```
:'-(

**Comment by nicolas-raoul** [comment]
If you update roboelectric to 4.10.3, the number of failed tests falls from all to only 5. :-)

**Comment by shankarpriyank** [comment]
hi @siva-subramaniam-v can you please push this fix, it would help a lot

> If you update roboelectric to 4.10.3, the number of failed tests falls from all to only 5. :-) It is OK to add an `@Ignore` to these 5 unit tests, and create an issue so that someone can fix them in the future.

**Comment by nicolas-raoul** [comment]
Now would be a great time to merge this.
Sorry for the delay, but would you mind rebasing (or pulling) from the latest main branch?
Thanks a lot!

**Comment by nicolas-raoul** [comment]
I successfully uploaded a picture using prodRelease of this branch:
https://commons.wikimedia.org/wiki/File:March_2007_night_in_Valetta_9.jpg

**Comment by nicolas-raoul** [comment]
The next task is to fix the unit tests, they all fail with this stacktrace:
```
 fr.free.nrw.commons.AboutActivityUnitTests > checkActivityNotNull FAILED
    org.mockito.exceptions.base.MockitoException at TestCommonsApplication.kt:81
        Caused by: org.mockito.exceptions.base.MockitoException at TypeCache.java:152
            Caused by: java.lang.IllegalStateException at InlineBytecodeGenerator.java:177
                Caused by: java.lang.IllegalArgumentException at ClassReader.java:184
```
Updating `jacoco.gradle`'s `toolVersion` to `0.8.10` does not fix them.
In `app/build.gradle` I see that we are using 2.* mockito packages, while the latest versions seem to be 5.*, time to update maybe?

How about releasing a beta right now even though the unit tests fail? By doing so, we give more time to testers while we fix the unit tests (thus allowing us to shorten a bit the final beta test phase).

**Comment by sivaraam** [comment]
> How about releasing a beta right now even though the unit tests fail? By doing so, we give more time to testers while we fix the unit tests (thus allowing us to shorten a bit the final beta test phase).

Ok Nicolas. Let us keep the changes of this branch in a separate release branch and release the beta from there. We could merge to `main` once we've fixed the unit tests. Right now the [`v4.2.0-release`](https://github.com/commons-app/apps-android-commons/tree/v4.2.0-release) has all the changes of this MR. Let's use the same 👍🏼

**Comment by nicolas-raoul** [comment]
Sounds good, thanks! :-)

**Comment by sivaraam** [comment]
Thanks a lot for contributing this change @siva-subramaniam-v ! It helped much for the release of the app 🙂

**Pull request merged**